[llvm][YAML] Support aliases (anchors) in YAMLIO

sizn-sc opened the pull request:
This change makes it possible to use YAML anchors [1], [2] with YAMLTraits. All of the necessary parser machinery already exists, so the only change that is necessary is to wire it up to YAMLTraits.

This is done by keeping track of all `Anchor` -> `HNode *` mappings and reusing those when an `AliasNode` is encountered.

In accordance with the spec [2], anchors do not have to be unique and refer to the last occurrence in the serialization.

Example usage:

```yaml
foo: &a 42
bar: *a
```

The above would be deserialized as:

```yaml
foo: 42
bar: 42
```

Note that aliases are a serialization detail and can be discarded during composition into a Representation Graph (`HNode` hierarchy).

[1]: https://yaml.org/spec/1.2.2/#692-node-anchors
[2]: https://yaml.org/spec/1.2.2/#3222-anchors-and-aliases

sizn-sc committed:
[llvm][YAML] Support aliases (anchors) in YAMLIO

This change makes it possible to use YAML anchors [1], [2]
with YAMLTraits. All of the necessary parser machinery already
exists, so the only change that is necessary is to wire it
up to YAMLTraits.

This is done by keeping track of all `Anchor` -> `HNode *`
mappings and reusing those when an `AliasNode` is encountered.

In accordance with the spec [2], anchors do not have to be unique and
refer to the last occurrence in the serialization.

Example usage:

```yaml
foo: &a 42
bar: *a
```

The above would be deserialized as:

```yaml
foo: 42
bar: 42
```

Note that aliases are a serialization detail and can be discarded during
composition into a Representation Graph (`HNode` hierarchy).

[1]: https://yaml.org/spec/1.2.2/#692-node-anchors
[2]: https://yaml.org/spec/1.2.2/#3222-anchors-and-aliases

atamazov commented:
@jasilvanus, @vitalybuka, @chapuni I am second to the above request

jh7370 reviewed (commented):
Just trying to wrap my head around this, which I think is mostly confused by the YAML terminology rather than anything else: am I right in thinking that this could be used with yaml2obj to, for example, simplify a YAML doc that contained a series of identical sections, so that you'd describe it in the first case, then reference the alias thereafter? Something like:

```
---!ELF
<snip>
Sections:
  - Name: .sec
    Type: SHT_PROGBITS
  - Name: .sec
    Type: SHT_PROGBITS
  - Name: .sec
    Type: SHT_PROGBITS
```
could become
```
---!ELF
<snip>
Sections:
  - &section
    Name: .sec
    Type: SHT_PROGBITS
  - *section
  - *section
```

jh7370 commented on llvm/lib/Support/YAMLTraits.cpp:
```diff
@@ -397,6 +397,17 @@ void Input::releaseHNodeBuffers() {
 }
 
+void Input::saveAliasHNode(Node *N, HNode *HN) {
+  if (auto Anchor = N->getAnchor(); !Anchor.empty())
```
I don't see a benefit to this syntax in this context (the variable is already scoped to the function, which is the same length as the if), so I'd prefer to put the assignment before the if.

Also, this isn't one of the cases where `auto` is appropriate (see the coding standards), so please use the actual type.

jh7370 reviewed (commented):
Please don't force push to changes under active review, as it trashes the incremental reviewing experience and so is against policy.

One nit, otherwise this looks good.

sizn-sc committed:
[llvm][YAML] Clean up saveAliasHNode

Addresses review comment https://github.com/llvm/llvm-project/pull/146371#discussion_r3115961233

sizn-sc commented:
@jh7370, thanks for the review and sorry for the force push. Pushed the style fix for https://github.com/llvm/llvm-project/pull/146371#discussion_r3115961233 as a separate commit now.

jh7370 reviewed (approved):
LGTM, assuming the pre-merge checks pass, but would be nice if @vitalybuka also approved before this gets merged. If they don't respond in a week, please ping me and I can merge it.

jh7370 commented:
The pre-merge check failure looks to be unrelated. You can merge changes from main via a merge commit (which will disappear when the PR gets squashed and merged), to bring it up to date. That'll hopefully fix the issue.

vitalybuka commented on llvm/unittests/Support/YAMLIOTest.cpp:
Error is find, some protection and test against infinit recursion would be nice

vitalybuka reviewed (approved):
It would be nice to implemented recursion protection
Other than that is LGTM.

sizn-sc committed:
Merge remote-tracking branch 'upstream/main' into yamlio-anchors

sizn-sc commented:
@jh7370, merged main to hopefully fix the CI. Could you kick off the pipelines again?

jh7370 commented:
> @jh7370, merged main to hopefully fix the CI. Could you kick off the pipelines again?

Done.

sizn-sc commented:
@jh7370, is this good to merge now?
